### PR TITLE
[DNM][WIP]Donator Items

### PR DESCRIPTION
**Pull request by francinum**
## About The Pull Request
At the behest of KMC.

## Why It's Good For The Game
At the behest of KMC.

## Changelog
:cl: Bezos2000
add: Donator items can now be bought through some means or something.
/:cl:

### CODE REVIEW

File: code/controllers/configuration/configuration.dm [hunk 1]
```diff
@@ -25,6 +25,8 @@
 
 	var/list/fail2topic_whitelisted_ips
 
+	var/list/active_donators
```
Comment (Jalleo) [inline code/controllers/configuration/configuration.dm:28]: Make sure to protect this list.

Comment (francinum) [reply code/controllers/configuration/configuration.dm]: There's a massive issue that prevents protecting lists.

Regardless, there's redundant measures in place that *in theory* should prevent people from both acquiring and using the donator items of other people, and if someone overrides this list it'll be logged and they'll get a massive paddling.

Unless you want to fix the list security flaw.

**Comment by francinum** [comment]
Part 2 PR after merge will handle modularization (Submodule fuckery)

File: code/modules/jobs/job_types/_job.dm [hunk 2]
```diff
@@ -86,18 +86,34 @@
...
+					to_chat(M, "<span class='warning'>Your current role does not permit you to spawn with [gear]!</span>")
 					permitted = FALSE
 
 				if(G.species_blacklist && (human.dna.species.id in G.species_blacklist))
+					to_chat(M, "<span class='warning'>Your current species does not permit you to spawn with [gear]!</span>")
 					permitted = FALSE
 
 				if(G.species_whitelist && !(human.dna.species.id in G.species_whitelist))
 					permitted = FALSE
+					if(G.unlocktype == GEAR_DONATOR) //Strip out donation items if the patreon has expired or they somehow have someone else's gear datum.
```
Comment (Kmc2000) [inline code/modules/jobs/job_types/_job.dm:98]: Should this not be at the top? This looks like it only runs if the gear has a species whitelist flag.

File: code/modules/client/preferences.dm [hunk 3]
```diff
@@ -1183,13 +1185,26 @@ GLOBAL_LIST_EMPTY(preferences_datums)
...
+					if(usr.ckey != TG.ckey)
+						//The fuck are you playing at?
+						log_href_exploit(usr)
+						return
+					if(usr.ckey in config.active_donators)
+						purchased_gear += TG.display_name
+						TG.purchase(user.client)
+						save_preferences()
+					else
+						to_chat(user,"<span class='warning'>Your donation has expired or not yet validated. Donation status is refreshed at round end.</span>")
```
Comment (Kmc2000) [inline code/modules/client/preferences.dm:1207]:
```suggestion
						to_chat(user,"<span class='warning'>Your donation has expired or has yet to be validated. Donation status is refreshed at round end.</span>")

```

File: code/modules/client/loadout/loadout_donator.dm [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+/datum/gear/purchase(client/C)
+	//This can be used to explain the code function of the item or otherwise cover ass.
+	to_chat(C, "<span class='boldwarning'>Your donator item has been added to your account. It will be removed automatically if your donation lapses.</span>")
+
+
+/datum/gear/donator/example
+	display_name = "Example Donator Item"
+	description = "Example Donator Description"
+	ckey = "francinum" //ckey should be in canonical format.
+	path = /obj/item/storage/toolbox/mechanical
```
Comment (Kmc2000) [inline code/modules/client/loadout/loadout_donator.dm:16]: Remember to remove this when you're done, or you'll be able to spawn with a toolbox :eyes:

File: code/controllers/configuration/configuration.dm (hunk 1 above)
Comment (Jalleo) [reply code/controllers/configuration/configuration.dm]: I have no idea how to fix that flaw.

File: code/modules/client/loadout/loadout.dm [hunk 5]
```diff
@@ -43,14 +45,16 @@ GLOBAL_LIST_EMPTY(gear_datums)
...
 /datum/gear
-	var/display_name       //Name/index. Must be unique.
-	var/description        //Description of this gear. If left blank will default to the description of the pathed item.
-	var/path               //Path to item.
-	var/cost = INFINITY           //Number of metacoins
-	var/slot               //Slot to equip to.
-	var/list/allowed_roles //Roles that can spawn with this item.
-	var/list/species_blacklist //Stop certain species from receiving this gear
-	var/list/species_whitelist //Only allow certain species to receive this gear
+	var/display_name               //Name/index. Must be unique.
```
Comment (Jalleo) [inline code/modules/client/loadout/loadout.dm:48]: why the hell does git think this is a new line?

File: code/modules/client/loadout/loadout_donator.dm (hunk 4 above)
Comment (Jalleo) [reply code/modules/client/loadout/loadout_donator.dm:16]: Change it to a fish for safety.

File: code/modules/jobs/job_types/_job.dm [hunk 6]
```diff
@@ -86,18 +86,34 @@
...
+							permitted = FALSE
+
+						if(!(M.client.ckey in config.active_donators))
+							to_chat(M, "<span class='warning'>Your patreon has expired! Your donator item has been removed. Item: [gear]</span>")
+							M.client.prefs.equipped_gear -= gear
+							M.client.prefs.purchased_gear -= gear
+							permitted = FALSE
 
 				if(!permitted)
-					to_chat(M, "<span class='warning'>Your current species or role does not permit you to spawn with [gear]!</span>")
```
Comment (Jalleo) [inline code/modules/jobs/job_types/_job.dm:98]: shouldn't that feedback?

File: code/controllers/configuration/configuration.dm (hunk 1 above)
Comment (francinum) [reply code/controllers/configuration/configuration.dm]: Then this is just going to have to remain unprotected.

File: code/modules/jobs/job_types/_job.dm (hunk 2 above)
Comment (francinum) [reply code/modules/jobs/job_types/_job.dm:98]: Patched. VSCODE is kinda weird with DM forced tabulation

File: code/modules/client/loadout/loadout.dm (hunk 5 above)
Comment (francinum) [reply code/modules/client/loadout/loadout.dm:48]: I messed with the spacing of the comments.

File: code/modules/jobs/job_types/_job.dm (hunk 6 above)
Comment (francinum) [reply code/modules/jobs/job_types/_job.dm:98]: Nah. each thing that would trigger this has been given it's own feedback.

**Pull request merged**